### PR TITLE
Prevent auth error from being reshown

### PR DESCRIPTION
**Pull request by rehandalal**
r?

### CODE REVIEW

**Comment by peterbe** [comment]
I struggle to understand this. But I'm not going to stand in its way if it works.

First of all, how do I reproduce this?

So what happens is that the props change, which causes a re-render, which causes a `componentDidUpdate` which would trigger calling `notification.error()` again. Is that right?
Then why use a timer and not just `if (authError && (!prevProps.authError || authError !== prevProps)) {`?

I'm guessing (and assuming) that the problem at hand is that the error notification growl-like message alert is displayed repeatedly with the same message. Is that right?

**Comment by rehandalal** [comment]
> First of all, how do I reproduce this?

1. Make sure you're on `master`.
2. Install [Redux DevTools](https://addons.mozilla.org/en-US/firefox/addon/remotedev/)
3. In Redux DevTools select the dispatcher and dispatch the following action:
   ```js
   {
     type: 'USER_AUTH_FAILURE',
     error: {
       code: 'some_error',
       description: 'OH NOES!'
     }
   }
   ```
4. Navigate around DC and you will see multiple notifications.

> So what happens is that the props change, which causes a re-render, which causes a `componentDidUpdate` which would trigger calling `notification.error()` again. Is that right?

Yep.

> Then why use a timer and not just if `(authError && (!prevProps.authError || authError !== prevProps)) {`?

Because if the same error occurs twice (e.g.: timeout and then timeout again) it will not show a notification for the second error because it thinks nothing has changed. The timer allows us to distinguish between the same type of error but different instances.

**Comment by peterbe** [comment]
Is there no better way to do this using rendering? E.g. 

```javascript
// In components.APP

render() {
  return (
      <ConnectedRouter history={this.props.history}>
        <Layout>
        <AuthError error={this.props.authError}/>
       ...
}
```
Then, since `authError` is an immutable Map instance, it would be smart enough to not update if the instance hasn't changed. 

Perhaps you can't call `notification.error` in a render. Just throwing around ideas.

**Comment by rehandalal** [comment]
>  it would be smart enough to not update if the instance hasn't changed.

I could try this. But I think the issue of the "same error" occurring multiple times is not solved by this.

I think another solution would be to use an action to show the notification and have it dispatch an action that clears the error once the notification is shown. Maybe that's a better way to do this.

**Comment by rehandalal** [comment]
> This is what happens when you wake up the computer after a long sleep.

Yup, this happens to me every morning!

**Comment by rehandalal** [comment]
@peterbe took a new approach that I think is cleaner.

File: src/components/App.js [hunk 1]
```diff
@@ -50,11 +56,7 @@ export default class App extends React.Component {
     }
 
     if (authError) {
-      notification.error({
-        message: 'Authentication Error',
-        description: `${authError.get('code')}: ${authError.get('description')}`,
-        duration: 0,
-      });
+      this.props.notifyAuthenticationError();
```
Comment (peterbe) [inline src/components/App.js]: `this.props.notifyAuthenticationError(authError)` surely?

File: src/state/auth/selectors.js [hunk 2]
```diff
@@ -11,7 +11,7 @@ export function getAccessToken(state, defaultsTo = null) {
  *
  * @param state  The current state
  * @param defaultsTo  A fallback value
- * @returns {Immutable.Map({code: String, description: String})}
+ * @returns {Immutable.Map({code: String, description: String, time: Integer})}
```
Comment (peterbe) [inline src/state/auth/selectors.js]: That extra `time` is no longer needed.

File: src/components/App.js (hunk 1 above)
Comment (rehandalal) [reply src/components/App.js]: oops yes! i've been waiting for my login to expire to verify this.

Comment (rehandalal) [reply src/components/App.js]: i guess i should just fake it with redux devtools. derp!

File: src/state/auth/selectors.js (hunk 2 above)
Comment (rehandalal) [reply src/state/auth/selectors.js]: good catch!

**Comment by rehandalal** [comment]
Verified with redux-devtools. Should work correctly now.

**Comment by peterbe** [comment]
bors r= @peterbe

**Pull request merged**